pkg/kubelet/util/pluginwatcher/example_handler.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pluginwatcher
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"golang.org/x/net/context"
+
+	v1beta1 "k8s.io/kubernetes/pkg/kubelet/util/pluginwatcher/example_plugin_apis/v1beta1"
+	v1beta2 "k8s.io/kubernetes/pkg/kubelet/util/pluginwatcher/example_plugin_apis/v1beta2"
+)
+
+type exampleHandler struct {
+	registeredPlugins        map[string]bool
+	mutex                    sync.Mutex
+	chanForHandlerAckTesting chan error // for testing
+}
+
+func NewExampleHandler() *exampleHandler {
+	return &exampleHandler{
+		chanForHandlerAckTesting: make(chan error),
+		registeredPlugins:        make(map[string]bool),
+	}
+}
+
+func (h *exampleHandler) Cleanup() error {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+	h.registeredPlugins = make(map[string]bool)
+	return nil
+}
+
+func (h *exampleHandler) Handler(pluginName string, endpoint string, versions []string, sockPath string) (error, chan bool) {
```

Comment:
Hi @figo,

@vladimirvivien has a point. The standard Go syntax is that the error should always be returned as the final value of a tuple/triple/etc. Additionally:

* `chan bool` should be `<-chan bool` as channels returned from a function should almost always be read-only. The channel inside the function can be `chan bool`, but returning `<-chan bool` prevents anything outside the function from closing or sending data into the channel. 
* Additionally, if you're returning a `<-chan T` as a result, you should ask yourself if the return signature should be: `(<-chan T, <-chan error, error)`. Remember, if you're returning results via a channel, there's no way to return an error while the first channel is still open. In other words, the caller **always** has to wait the full second (from the code below) to determine if the function returns an error or not.

Reply:
i agree that that error usually return last, i will update, thanks.

in term of channel, this is not a read-only channel, caller need to write it.

Reply:
Hi @figo,

Re: error. It’s not a matter of agreement. If you enable Vim-Go or use an IDE with a Go plug-in, any Go linter is going to call out the function signature as an error. It would be like making context.Context anything but the first parameter in a function.

Also, if the caller is writing to the channel then you should be passing the channel into the function. As is there is a design bug where this function closes a channel and thus a caller can write to a closed channel causing a panic.

What’s the purpose of the channel?

Reply:
for the channel design question, it need be discussed separately, it probably be addressed by:
https://github.com/kubernetes/kubernetes/pull/64621, @RenaudWasTaken 

this patch does not change the channel design, so it is a writable channel.